base64ct/src/lib.rs

```diff
@@ -364,44 +355,66 @@ fn decode_3bytes(src: &[u8], dst: &mut [u8]) -> isize {
 }
 
 #[inline(always)]
-fn decode_3bytes_padded(src: &[u8], dst: &mut [u8]) -> isize {
-    let mut i = 0;
+fn decode_3bytes_padded(src: &[u8], dst: &mut [u8]) -> i16 {
     let mut err = 0;
     let mut tmp_out = [0u8; 3];
     let mut tmp_in = [b'A'; 4];
 
-    while i < src.len() && src[i] != PAD {
-        tmp_in[i] = src[i];
-        i += 1;
-    }
+    let src_len = unpadded_len_ct(src);
 
-    if i < 2 {
+    if src_len < 2 {
         err = 1;
     }
 
-    let len = i - 1;
+    tmp_in[..src_len].copy_from_slice(&src[..src_len]);
+
+    let dst_len = src_len - 1;
     err |= decode_3bytes(&tmp_in, &mut tmp_out);
-    dst[..len].copy_from_slice(&tmp_out[..len]);
+    dst[..dst_len].copy_from_slice(&tmp_out[..dst_len]);
     err
 }
 
 #[inline(always)]
-fn decode_6bits(src: u8) -> isize {
-    let ch = src as isize;
-    let mut ret: isize = -1;
+fn decode_6bits(src: u8) -> i16 {
+    let mut res: i16 = -1;
+
+    // if (byte > 0x40 && byte < 0x5b) res += byte - 0x41 + 1; // -64
+    res += match_byte_range_ct(src, 0x40, 0x5b, src as i16 - 64);
+
+    // if (byte > 0x60 && byte < 0x7b) res += byte - 0x61 + 26 + 1; // -70
+    res += match_byte_range_ct(src, 0x60, 0x7b, src as i16 - 70);
+
+    // if (byte > 0x2f && byte < 0x3a) res += byte - 0x30 + 52 + 1; // 5
+    res += match_byte_range_ct(src, 0x2f, 0x3a, src as i16 + 5);
+
+    // if (byte == 0x2b) res += 62 + 1;
+    res += match_byte_ct(src, 0x2b, 63);
+
+    // if (byte == 0x2f) res += 63 + 1;
+    res + match_byte_ct(src, 0x2f, 64)
+}
 
-    // if (ch > 0x40 && ch < 0x5b) ret += ch - 0x41 + 1; // -64
-    ret += (((64isize - ch) & (ch - 91isize)) >> 8) & (ch - 64isize);
+/// Pseudo-branch operation
+#[inline(always)]
+fn match_byte_range_ct(input: u8, lo: u8, hi: u8, ret_on_match: i16) -> i16 {
+    (((lo as i16 - input as i16) & (input as i16 - hi as i16)) >> 8) & ret_on_match
+}
 
-    // if (ch > 0x60 && ch < 0x7b) ret += ch - 0x61 + 26 + 1; // -70
-    ret += (((96isize - ch) & (ch - 123isize)) >> 8) & (ch - 70isize);
+/// Match a specific byte value
+#[inline(always)]
+fn match_byte_ct(input: u8, expected: u8, ret_on_match: i16) -> i16 {
+    match_byte_range_ct(input, expected - 1, expected + 1, ret_on_match)
+}
 
-    // if (ch > 0x2f && ch < 0x3a) ret += ch - 0x30 + 52 + 1; // 5
-    ret += (((47isize - ch) & (ch - 58isize)) >> 8) & (ch + 5isize);
+/// Compute the length of the unpadded portion of a Base64-encoded string
+/// in constant time
+fn unpadded_len_ct(input: &[u8]) -> usize {
+    let mut res = 0;
 
-    // if (ch == 0x2b) ret += 62 + 1;
-    ret += (((42isize - ch) & (ch - 44isize)) >> 8) & 63;
+    // TODO(tarcieri): ensure padding bytes are always at the end of a string
+    for &byte in input {
+        res += match_byte_ct(byte, PAD, 1) as usize ^ 1;
+    }
```

Comment:
You iterate over the whole string here, which is not ideal. I think it would be better to check if `input` length is larger or equal to 4, and then subtract from the length if last bytes match `PAD`.

Reply:
Sure, that's a better approach

Reply:
Replaced it with subslice matching, which avoids the need to do any length calculations.

[Godbolt output](https://godbolt.org/z/sW6MGT)